csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -452,7 +452,17 @@ public void WriteTo(CodedOutputStream output, Codec codec)
             WriteContext.Initialize(output, out WriteContext ctx);
             try
             {
-                WriteTo(ref ctx, codec);
+                if (output.IsSerializationDeterministic())
+                {
+                    var sorted = new List<KeyValuePair<TKey, TValue>>(list);
+                    sorted.Sort((pair1, pair2) => Comparer<TKey>.Default.Compare(pair1.Key, pair2.Key));
+
+                    WriteTo(ref ctx, codec, sorted);
+                }
+                else
+                {
+                    WriteTo(ref ctx, codec, list);
```

Comment:
I think it would be better not to duplicate the `WriteTo` call here. Something like:

```csharp
IList<KeyValuePair<TKey, TValue>> listToWrite = list;
if (output.IsDeterministic)
{
    listToWrite = new List<KeyValuePair<TKey, TValue>>(list);
    listToWrite.Sort(...);
}
WriteTo(ref ctx, codec, listToWrite);
```

Reply:
refactored to call `GetSortedListCopy`